File: web/app/services/transaction_service.py
"""Money movement: deposits, withdrawals, transfers and transaction history."""
from .. import money
from ..database import get_db
from .account_service import get_account

TYPE_MAP = {
    "DEPOSIT": "Deposit",
    "WITHDRAW": "Withdrawal",
    "TRANSFER_IN": "Transfer in",
    "TRANSFER_OUT": "Transfer out",
    "OPENING": "Opening",
    "INTEREST": "Interest",
    "CLOSED": "Closed",
}


def _record(db, number, tx_type, amount_paise, balance_after_paise,
            counterparty, remark):
    db.execute(
        "INSERT INTO transactions (account_number, type, amount_paise,"
        " balance_after_paise, counterparty, timestamp, remark)"
        " VALUES (?, ?, ?, ?, ?, ?, ?)",
        (number, tx_type, amount_paise, balance_after_paise,
         counterparty, money.now(), remark),
    )


def deposit(number, paise, counterparty="Cash deposit", remark="Deposit"):
    """Deposit into an account. Returns error or None."""
    if paise <= 0:
        return "Amount must be positive."
    db = get_db()
    acct = get_account(number)
    if acct is None:
        return "Account not found."
    if acct["status"] != "Active":
        return "Account is closed — deposits are not allowed."
    new_balance = acct["balance_paise"] + paise
    db.execute(
        "UPDATE accounts SET balance_paise = ? WHERE account_number = ?",
        (new_balance, number),
    )
    _record(db, number, "DEPOSIT", paise, new_balance, counterparty, remark)
    db.commit()
    return None


def withdraw(number, paise, counterparty="Cash withdrawal", remark="Withdrawal"):
    """Withdraw from an account (overdrafts not allowed). Returns error or None."""
    if paise <= 0:
        return "Amount must be positive."
    db = get_db()
    acct = get_account(number)
    if acct is None:
        return "Account not found."
    if acct["status"] != "Active":
        return "Account is closed — withdrawals are not allowed."
    if paise > acct["balance_paise"]:
        return f"Insufficient balance. Available: {money.format_paise(acct['balance_paise'])}"
    new_balance = acct["balance_paise"] - paise
    db.execute(
        "UPDATE accounts SET balance_paise = ? WHERE account_number = ?",
        (new_balance, number),
    )
    _record(db, number, "WITHDRAW", paise, new_balance, counterparty, remark)
    db.commit()
    return None


def transfer(from_number, to_number, paise, remark="Transfer"):
    """Move money between two accounts. Returns error or None."""
    if paise <= 0:
        return "Amount must be positive."
    if from_number == to_number:
        return "Source and destination accounts must be different."
    db = get_db()
    src = get_account(from_number)
    dst = get_account(to_number)
    if src is None or dst is None:
        return "One of the accounts was not found."
    if src["status"] != "Active":
        return "Source account is closed."
    if dst["status"] != "Active":
        return "Destination account is closed."
    if paise > src["balance_paise"]:
        return f"Insufficient balance. Available: {money.format_paise(src['balance_paise'])}"

    src_name = db.execute("SELECT name FROM users WHERE id = ?", (src["user_id"],)).fetchone()
    dst_name = db.execute("SELECT name FROM users WHERE id = ?", (dst["user_id"],)).fetchone()
    src_label = src_name["name"] if src_name else ""
    dst_label = dst_name["name"] if dst_name else ""

    new_src = src["balance_paise"] - paise
    new_dst = dst["balance_paise"] + paise
    db.execute(
        "UPDATE accounts SET balance_paise = ? WHERE account_number = ?",
        (new_src, from_number),
    )
    db.execute(
        "UPDATE accounts SET balance_paise = ? WHERE account_number = ?",
        (new_dst, to_number),
    )
    _record(db, from_number, "TRANSFER_OUT", paise, new_src, to_number, f"{remark} to {dst_label}")
    _record(db, to_number, "TRANSFER_IN", paise, new_dst, from_number, f"{remark} from {src_label}")
    db.commit()
    return None


def history(number, limit=100):
    db = get_db()
    return db.execute(
        "SELECT * FROM transactions WHERE account_number = ?"
        " ORDER BY id DESC LIMIT ?",
        (number, limit),
    ).fetchall()


def all_history(limit=200):
    db = get_db()
    return db.execute(
        "SELECT t.*, a.user_id FROM transactions t"
        " JOIN accounts a ON a.account_number = t.account_number"
        " ORDER BY t.id DESC LIMIT ?",
        (limit,),
    ).fetchall()


def today_activity():
    """Aggregate transaction counts and volume for today (admin dashboard)."""
    db = get_db()
    row = db.execute(
        "SELECT COUNT(*) AS count, COALESCE(SUM(amount_paise), 0) AS total"
        " FROM transactions WHERE timestamp LIKE ?",
        (money.today() + "%",),
    ).fetchone()
    return {"count": row["count"], "total_paise": row["total"]}


def last_seven_days():
    """Per-day transaction volume for the last 7 days (charts)."""
    db = get_db()
    from datetime import datetime, timedelta

    today_dt = datetime.now()
    days = []
    for i in range(6, -1, -1):
        d = today_dt - timedelta(days=i)
        prefix = d.strftime("%Y-%m-%d") + "%"
        row = db.execute(
            "SELECT COUNT(*) AS count, COALESCE(SUM(amount_paise), 0) AS total"
            " FROM transactions WHERE timestamp LIKE ?",
            (prefix,),
        ).fetchone()
        days.append({
            "date": d.strftime("%d %b"),
            "count": row["count"],
            "total_paise": row["total"],
        })
    return days

File: src/services/TransactionService.cpp
#include "services/TransactionService.h"

#include "core/Utils.h"
#include "models/Account.h"

namespace bms {

bool TransactionService::deposit(const std::string& accountNumber,
                                 long long amountPaise,
                                 const std::string& remark,
                                 std::string& err) {
    Account* acc = db_.findAccountByNumber(accountNumber);
    if (!acc) {
        err = "Account not found.";
        return false;
    }
    if (acc->status != AccountStatus::ACTIVE) {
        err = "Account is closed.";
        return false;
    }
    if (amountPaise <= 0) {
        err = "Amount must be positive.";
        return false;
    }

    acc->balancePaise += amountPaise;

    Transaction tx;
    tx.id = db_.nextTransactionId();
    tx.accountNumber = accountNumber;
    tx.type = TxType::DEPOSIT;
    tx.amountPaise = amountPaise;
    tx.balanceAfterPaise = acc->balancePaise;
    tx.counterparty = "Bank";
    tx.timestamp = Utils::now();
    tx.remark = remark.empty() ? "Cash deposit" : remark;
    db_.addTransaction(tx);
    db_.saveAll();
    return true;
}

bool TransactionService::withdraw(const std::string& accountNumber,
                                  long long amountPaise,
                                  const std::string& remark,
                                  std::string& err) {
    Account* acc = db_.findAccountByNumber(accountNumber);
    if (!acc) {
        err = "Account not found.";
        return false;
    }
    if (acc->status != AccountStatus::ACTIVE) {
        err = "Account is closed.";
        return false;
    }
    if (amountPaise <= 0) {
        err = "Amount must be positive.";
        return false;
    }
    if (!checkSufficientBalance(*acc, amountPaise)) {
        err = "Insufficient balance.";
        return false;
    }

    acc->balancePaise -= amountPaise;

    Transaction tx;
    tx.id = db_.nextTransactionId();
    tx.accountNumber = accountNumber;
    tx.type = TxType::WITHDRAWAL;
    tx.amountPaise = amountPaise;
    tx.balanceAfterPaise = acc->balancePaise;
    tx.counterparty = "Bank";
    tx.timestamp = Utils::now();
    tx.remark = remark.empty() ? "Cash withdrawal" : remark;
    db_.addTransaction(tx);
    db_.saveAll();
    return true;
}

bool TransactionService::transfer(const std::string& fromNumber,
                                  const std::string& toNumber,
                                  long long amountPaise,
                                  const std::string& remark,
                                  std::string& err) {
    if (fromNumber == toNumber) {
        err = "Cannot transfer to the same account.";
        return false;
    }
    Account* from = db_.findAccountByNumber(fromNumber);
    if (!from || from->status != AccountStatus::ACTIVE) {
        err = "Source account is not active.";
        return false;
    }
    Account* to = db_.findAccountByNumber(toNumber);
    if (!to || to->status != AccountStatus::ACTIVE) {
        err = "Destination account is not active.";
        return false;
    }
    if (amountPaise <= 0) {
        err = "Amount must be positive.";
        return false;
    }
    if (!checkSufficientBalance(*from, amountPaise)) {
        err = "Insufficient balance in the source account.";
        return false;
    }

    from->balancePaise -= amountPaise;
    to->balancePaise += amountPaise;

    const std::string ts = Utils::now();

    Transaction out;
    out.id = db_.nextTransactionId();
    out.accountNumber = fromNumber;
    out.type = TxType::TRANSFER_OUT;
    out.amountPaise = amountPaise;
    out.balanceAfterPaise = from->balancePaise;
    out.counterparty = toNumber;
    out.timestamp = ts;
    out.remark = remark.empty() ? "Funds transfer" : remark;
    db_.addTransaction(out);

    Transaction in;
    in.id = db_.nextTransactionId();
    in.accountNumber = toNumber;
    in.type = TxType::TRANSFER_IN;
    in.amountPaise = amountPaise;
    in.balanceAfterPaise = to->balancePaise;
    in.counterparty = fromNumber;
    in.timestamp = ts;
    in.remark = remark.empty() ? "Funds transfer" : remark;
    db_.addTransaction(in);

    db_.saveAll();
    return true;
}

std::vector<const Transaction*> TransactionService::statementOf(
    const std::string& accountNumber, int limit) const {
    std::vector<const Transaction*> out;
    for (const auto& tx : db_.transactions()) {
        if (tx.accountNumber == accountNumber) out.push_back(&tx);
    }
    if (static_cast<int>(out.size()) > limit) {
        out.erase(out.begin(), out.begin() + (out.size() - limit));
    }
    return out;
}

long long TransactionService::todayVolumePaise() const {
    const std::string today = Utils::today();
    long long volume = 0;
    for (const auto& tx : db_.transactions()) {
        if (tx.timestamp.compare(0, today.size(), today) == 0) {
            volume += tx.amountPaise;
        }
    }
    return volume;
}

bool TransactionService::checkSufficientBalance(const Account& acc,
                                                long long amountPaise) const {
    return acc.balancePaise >= amountPaise;
}

}  // namespace bms
